src/ProjectTemplates/test/SpaTemplateTest/SpaTemplateTestBase.cs

```diff
@@ -46,25 +53,84 @@ protected void SpaTemplateImpl(string template, bool noHttps = false)
+        // it's important to distribute the SPA template tests over different test classes
+        // so they can be run in parallel. Xunit doesn't parallelize within a test class.
+        protected async Task SpaTemplateImpl_IndividualAuthAsync(string key, string template, bool useLocalDb = false, bool noHttps = false)
+        {
+            Project = ProjectFactory.CreateProject(key, Output);
+
+            Project.RunDotNetNew(template, auth: "Individual", language: null, useLocalDb, noHttps: noHttps);
+
+            // For some SPA templates, the NPM root directory is './ClientApp'. In other
+            // templates it's at the project root. Strictly speaking we shouldn't have
+            // to do the NPM restore in tests because it should happen automatically at
+            // build time, but by doing it up front we can avoid having multiple NPM
+            // installs run concurrently which otherwise causes errors when tests run
+            // in parallel.
+            var clientAppSubdirPath = Path.Combine(Project.TemplateOutputDir, "ClientApp");
+            Assert.True(File.Exists(Path.Combine(clientAppSubdirPath, "package.json")), "Missing a package.json");
+
+            Npm.RestoreWithRetry(Output, clientAppSubdirPath);
+            Npm.Test(Output, clientAppSubdirPath);
+
+            var projectFileContents = Project.ReadFile($"{Project.ProjectName}.csproj");
+            if (!useLocalDb)
+            {
+                Assert.Contains(".db", projectFileContents);
+            }
+
+            Project.RunDotNetEfCreateMigration(template);
+
+            Project.AssertEmptyMigration(template);
+
+
+            await TestApplicationAsync(publish: false, visitFetchData: false);
+
+            UpdateSettingsForPublish();
+
+            await TestApplicationAsync(publish: true, visitFetchData: false);
+        }
+
+        private void UpdateSettingsForPublish()
+        {
+            // Hijack here the config file to use the development key during publish.
+            var appSettings = JObject.Parse(File.ReadAllText(Path.Combine(Project.TemplateOutputDir, "appsettings.json")));
+            var appSettingsDevelopment = JObject.Parse(File.ReadAllText(Path.Combine(Project.TemplateOutputDir, "appsettings.Development.json")));
+            ((JObject)appSettings["IdentityServer"]).Merge(appSettingsDevelopment["IdentityServer"]);
+            ((JObject)appSettings["IdentityServer"]).Merge(new
+            {
+                IdentityServer = new
+                {
+                    Key = new
+                    {
+                        FilePath = "./tempkey.json"
+                    }
+                }
+            });
+            var testAppSettings = appSettings.ToString();
+            File.WriteAllText(Path.Combine(Project.TemplateOutputDir, "appsettings.json"), testAppSettings);
+        }
+
+        private async Task TestApplicationAsync(bool publish, bool visitFetchData = true)
```

Comment:
Why wouldn't we visit fetchData?

Reply:
Because it requires a different flow than in the unauthenticated flow and requires templates changes that are not in yet.